File: memcached-itool.php
<?php
/**
 * @author: Andrey Niakhaichyk andrey@niakhaichyk.org
 * @version: 1.0
 */

const DEFAULT_TCP_TIMEOUT = 30;
const DEFAULT_UNIXSOCKET_TIMEOUT = 5;

const DUMPMODE_ONLYKEYS = 0;
const DUMPMODE_KEYVALUES = 1;
const REMOVEMODE_EXPIRED = 2;

$params = $_SERVER['argv'];

// Parse command line
$host = 'localhost';
$port = 11211;
$socketpath = NULL;
$mode = 'display';
if(!empty($params[1]))
{
  if(strpos($params[1], '/') === 0)
  {
    $socketpath = $params[1];
    $host = NULL;
    $port = NULL;
  }
  elseif(strpos($params[1], ':') > 0)
    list($host, $port) = explode(':', $params[1]);
  else
    $host = $params[1];
}
if(!empty($params[2]))
{
  $mode = $params[2];
}

// Check params
if(count($params) < 2 || !in_array($mode, array('display', 'dumpkeys', 'dump', 'removeexp', 'settings', 'stats', 'sizes')))
{
  help();
  exit;
}

// Connect to memcached
$errno = NULL;
$errstr = NULL;
if($host && $port)
  $fp = stream_socket_client("tcp://{$host}:{$port}", $errno, $errstr, DEFAULT_TCP_TIMEOUT);
else
  $fp = stream_socket_client("unix://{$socketpath}", $errno, $errstr, DEFAULT_UNIXSOCKET_TIMEOUT);

if (!$fp)
{
  echo "$errstr ($errno)".PHP_EOL;
  exit -1;
}


// Run logic
switch($mode)
{
  case 'settings':
    show_stats($fp, 'stats settings');
    break;

  case 'stats':
    show_stats($fp, 'stats');
    break;

  case 'sizes':
    sizes($fp);
    break;

  case 'dumpkeys':
    iterkeys($fp, DUMPMODE_ONLYKEYS);
    break;

  case 'removeexp':
    iterkeys($fp, REMOVEMODE_EXPIRED);
    break;

  case 'dump':
    iterkeys($fp, DUMPMODE_KEYVALUES);
    break;

  case 'display':
  default:
    display($fp);
    break;
}

exit;


// # -------------------- functions
function help()
{
  echo <<<HELP
 Usage: memcached-itool <host[:port] | /path/to/socket> [mode]

	memcached-itool localhost:11211 display    # shows slabs information (display is default mode)
	memcached-itool localhost:11211 dumpkeys   # dumps only keys names
	memcached-itool localhost:11211 dump       # dumps keys and values, values only for non expired keys
	memcached-itool localhost:11211 removeexp  # remove expired keys (you may need run several times)
	memcached-itool localhost:11211 settings   # shows memcached settings
	memcached-itool localhost:11211 sizes      # group keys by sizes and show how many we waste memory
	memcached-itool localhost:11211 stats      # shows general stats

Warning! dumpkeys, dump, removeexp and sizes modes *will* lock up your cache! It iterates over *every item* and examines the size. 
While the operation is fast, if you have many items you could prevent memcached from serving requests for several seconds.

Warning! dump and removeexp modes influence on memcached internal statistic like *expired_unfetched* and *get_misses*. So we recommend only use it for debugging purposes.

HELP;
}


function send_and_receive($fp, $command)
{
  fwrite($fp, $command."\r\n");
  $result = '';
  while (!feof($fp))
  {
    $result .= fgets($fp);

    if(strpos($result, 'END'."\r\n") !== FALSE)
      break;
  }
  
  $lines = explode("\r\n", $result);
  foreach($lines as $key=>$line)
  {
    if(strlen($line) == 0 || trim($line) == 'END')
      unset($lines[$key]);
  }

  return $lines;
}


function slabs_stats($fp)
{
  $slabs = array();

  $lines = send_and_receive($fp, 'stats slabs');
  foreach($lines as $line)
  {
    $m = array();
    if(preg_match('/^STAT (\d+):(\w+) (\d+)/', $line, $m))
    {
      $slab_num = $m[1];
      $slab_property = $m[2];
      $slab_value = $m[3];

      $slabs[$slab_num][$slab_property] = $slab_value;
    }

    if(preg_match('/^STAT (\w+) (\d+)/', $line, $m))
    {
      $slab_property = $m[1];
      $slab_value = $m[2];

      $slabs['total'][$slab_property] = $slab_value;
    }
  }

  $lines = send_and_receive($fp, 'stats items');
  foreach($lines as $line)
  {
    $m = array();
    if(preg_match('/^STAT items:(\d+):(\w+) (\d+)/', $line, $m))
    {
      $slab_num = $m[1];
      $slab_property = $m[2];
      $slab_value = $m[3];

      $slabs[$slab_num][$slab_property] = $slab_value;
    }
  }

  foreach($slabs as $num => $slab) 
  {
    if($num == 'total')
      continue;

    $slab['age'] = !empty($slab['age']) ? $slab['age'] : 0;
    $slab['number'] = !empty($slab['number']) ? $slab['number'] : 0;
    $slab['evicted'] = !empty($slab['evicted']) ? $slab['evicted'] : 0;
    $slab['evicted_time'] = !empty($slab['evicted_time']) ? $slab['evicted_time'] : 0;
    $slab['outofmemory'] = !empty($slab['outofmemory']) ? $slab['outofmemory'] : 0;

    $slabs[$num] = $slab;
  }
  ksort($slabs);

  return $slabs;
}


function get_stats($fp, $command = 'stats')
{
  $stats = array();

  $lines = send_and_receive($fp, $command);
  foreach($lines as $line)
  {
    $m = array();
    if(preg_match('/^STAT ([^\s]+) ([^\s]+)/', $line, $m))
    {
      $property = $m[1];
      $value = $m[2];

      $stats[$property] = $value;
    }
  }
  ksort($stats);

  return $stats;
}


function settings_stats($fp)
{
  $stats = array();

  $lines = send_and_receive($fp, 'stats settings');
  foreach($lines as $line)
  {
    $m = array();
    if(preg_match('/^STAT ([^\s]+) ([^\s]+)/', $line, $m))
    {
      $property = $m[1];
      $value = $m[2];

      $stats[$property] = $value;
    }
  }

  return $stats;
}


function display($fp)
{
  $slabs = slabs_stats($fp);

  print "  # Chunk_Size  Max_age   Pages   Count   Full?  Evicted Evict_Time OOM     Used   Wasted".PHP_EOL;
  foreach($slabs as $num => $slab) 
  {
    if($num == 'total')
      continue;

    $is_slab_full = $slab['free_chunks_end'] == 0 ? "yes" : "no";
    $wasted = $slab['number'] ? (1.0 - (float)$slab['mem_requested'] / ($slab['chunk_size'] * $slab['number'])) * 100 : 0.0;

    printf("%3d %10s %7ds %7d %7d %7s %8d %10d %3d %8s %7d%%".PHP_EOL, $num, descriptive_size($slab['chunk_size']), $slab['age'], $slab['total_pages'], $slab['number'], $is_slab_full, 
      $slab['evicted'], $slab['evicted_time'], $slab['outofmemory'], descriptive_size($slab['mem_requested']), $wasted);
  }

  print PHP_EOL."Total:".PHP_EOL;
  foreach($slabs['total'] as $property=>$value)
  {
    if($property == 'total_malloced')
      printf("%-15s %12s".PHP_EOL, $property, descriptive_size($value));
    else
      printf("%-15s %12s".PHP_EOL, $property, $value);
  }

  $stats = settings_stats($fp);
  $pages = 1;
  for($chunk_size = 96; $chunk_size * $stats['growth_factor'] < $stats['item_size_max']; $chunk_size *= $stats['growth_factor'])
    $pages++;
  printf("%-15s %12s (real %s - %s)".PHP_EOL, 'maxbytes', descriptive_size($stats['maxbytes']), descriptive_size(max($stats['item_size_max'] * $pages, $stats['maxbytes'])), descriptive_size($stats['item_size_max'] * ($pages + $stats['maxbytes'] / $stats['item_size_max'] - 1)));
  
  printf("%-15s %12s".PHP_EOL, 'item_size_max', descriptive_size($stats['item_size_max']));
  printf("%-15s %12s".PHP_EOL, 'evictions', $stats['evictions']);
  printf("%-15s %12s".PHP_EOL, 'growth_factor', $stats['growth_factor']);

}


function show_stats($fp, $command = 'stats')
{
  $stats = get_stats($fp, $command);
  ksort($stats);

  printf ("%24s %15s".PHP_EOL, "Field", "Value");
  foreach($stats as $property => $value)
    printf ("%24s %15s".PHP_EOL, $property, $value);
}


function iterkeys($fp, $dumpmode = DUMPMODE_ONLYKEYS)
{
  $slabs = slabs_stats($fp);
  $stats = get_stats($fp, 'stats');
  ksort($slabs);

  printf("      %-40s %20s %10s %8s".PHP_EOL, 'Key', 'Expire status', 'Size', 'Waste');
  foreach($slabs as $num => $slab) 
  {
    if($num == 'total')
      continue;

    if($slab['number'])
    {
      $lines = send_and_receive($fp, "stats cachedump {$num} {$slab['number']}");
      foreach($lines as $line)
      {
        $m = array();
        if(preg_match('/^ITEM ([^\s]+) \[(\d+) b; (\d+) s\]/', $line, $m))
        {
          $key = $m[1];
          $size = $m[2];

          $now = time();

          $waste = (1.0 - (float)$size / $slab['chunk_size']) * 100;
          $expiration_time = $m[3];
          if($expiration_time == $stats['time'] - $stats['uptime'])
            $status = '[never expire]';
          elseif($now > $expiration_time)
            $status = '[expired]';
          else
            $status = ($expiration_time - $now).'s left';

          printf("ITEM  %-40s %20s %10s %7.0f%%".PHP_EOL, $key, $status, $size, $waste);

          // Get value
          if($dumpmode == DUMPMODE_KEYVALUES && $status != '[expired]')
          {
            $lines = send_and_receive($fp, "get {$key}");
            if(count($lines))
            {
              $info = $lines[0];
              $data = $lines[1];
              preg_match('/^VALUE ([^\s]+) (\d+) (\d+)/', $info, $m);
              $flags = $m[2];
              printf("VALUE %-40s flags=%X".PHP_EOL, $key, $flags);
              printf("%s".PHP_EOL, $data);
            }
          }

          // Get value
          if($dumpmode == REMOVEMODE_EXPIRED && $status == '[expired]')
          {
            $lines = send_and_receive($fp, "get {$key}");
          }
        }
      }
    }
  }
}


function sizes($fp)
{
  $sizes = array();
  $stats = settings_stats($fp);

  printf("%-10s %10s %10s %10s".PHP_EOL, 'Size', 'Items', 'Chunk_Size', 'Wasted');

  $lines = send_and_receive($fp, 'stats sizes');
  foreach($lines as $line)
  {
    $m = array();
    if(preg_match('/^STAT ([^\s]+) ([^\s]+)/', $line, $m))
    {
      $size = $m[1];
      $values = $m[2];

      for($chunk_size = 96; $chunk_size * $stats['growth_factor'] < $size; $chunk_size *= $stats['growth_factor']);
      $chunk_size *= $stats['growth_factor'];
      if($chunk_size * $stats['growth_factor'] > $stats['item_size_max'])
        $chunk_size = $stats['item_size_max'];

      $wasted = (1.0 - $size / $chunk_size) * 100;

      printf("%-10s %10d %10s %9.0f%%".PHP_EOL, descriptive_size($size), $values, descriptive_size((int)$chunk_size), $wasted);

      $sizes[$size] = $values;
    }
  }
}


function descriptive_size($size)
{
  if($size >= 1024*1024)
    return sprintf('%.1fM', (float)$size/(1024*1024));
  if($size >= 1024)
    return sprintf('%.1fK', (float)$size/(1024));
  return $size.'B';
}

File: memcached-itool.py
# @author:  Andrey Niakhaichyk andrey@niakhaichyk.org
# @version: 1.0

import sys
import socket
import re
import time

DEFAULT_TCP_TIMEOUT = 30
DEFAULT_UNIXSOCKET_TIMEOUT = 5

DUMPMODE_ONLYKEYS = 0
DUMPMODE_KEYVALUES = 1
REMOVEMODE_EXPIRED = 2


def myhelp():
    print """
Usage: memcached-itool <host[:port] | /path/to/socket> [mode]

Modes:
  display    # shows slabs information (display is default mode)
  dumpkeys   # dumps only keys names
  dump       # dumps keys and values, values only for non expired keys
  removeexp  # remove expired keys (you may need run several times)
  settings   # shows memcached settings
  sizes      # group keys by sizes and show how many we waste memory
  stats      # shows general stats

Warning! dumpkeys, dump, removeexp and sizes modes *will* lock up your cache!
It iterates over *every item* and examines the size.
While the operation is fast, if you have many items you could prevent memcached
from serving requests for several seconds.

Warning! dump and removeexp modes influence on memcached internal statistic
like *expired_unfetched* and *get_misses*.
So we recommend only use it for debugging purposes.
"""
    return


def send_and_receive(sp, command):
    sp.send(command + '\r\n')

    result = ""
    while True:
        result += sp.recv(1024)
        if result.find("END" + '\r\n') >= 0:
            break

    lines = result.strip().split('\r\n')
    return lines[:-1]


def slabs_stats(sp):
    slabs = {'total': {}}

    lines = send_and_receive(sp, 'stats slabs')
    for line in lines:
        properties = re.match('^STAT (\d+):(\w+) (\d+)', line)
        if properties:
            slab_num, slab_property, slab_value = properties.groups()
            slab_num = int(slab_num)
            slab_value = int(slab_value)
            if slab_num not in slabs:
                slabs[slab_num] = {}
            slabs[slab_num][slab_property] = slab_value
            continue

        statistic = re.match('^STAT (\w+) (\d+)', line)
        if statistic:
            stat_property, stat_value = statistic.groups()
            slabs['total'][stat_property] = int(stat_value)

        pass

    lines = send_and_receive(sp, 'stats items')
    for line in lines:
        items_stat = re.match('^STAT items:(\d+):(\w+) (\d+)', line)
        if items_stat:
            slab_num, slab_property, slab_value = items_stat.groups()
            slabs[int(slab_num)][slab_property] = int(slab_value)
            continue

        pass

    for num, s in slabs.iteritems():
        if num == 'total':
            continue

        s['age'] = s['age'] if 'age' in s else 0
        s['number'] = s['number'] if 'number' in s else 0
        s['evicted'] = s['evicted'] if 'evicted' in s else 0
        s['evicted_time'] = s['evicted_time'] if 'evicted_time' in s else 0
        s['outofmemory'] = s['outofmemory'] if 'outofmemory' in s else 0

    return slabs


def descriptive_size(size):
    if size >= 1024*1024:
        return '{0:.1f}M'.format(float(size) / (1024*1024))
    if size >= 1024:
        return '{0:.1f}K'.format(float(size) / 1024)

    return str(int(size))+'B'


def get_stats(sp, command='stats'):
    stats = {}

    lines = send_and_receive(sp, command)
    for line in lines:
        items_stat = re.match('^STAT ([^\s]+) ([^\s]+)', line)
        if items_stat:
            sett_property, sett_value = items_stat.groups()
            stats[sett_property] = sett_value

    return stats


def show_stats(sp, command='stats'):
    stats = get_stats(sp, command)

    print "{0:>24s} {1:>15s}".format("Field", "Value")
    for prop in sorted(stats.keys()):
        print "{0:>24s} {1:>15s}".format(prop, stats[prop])

    return


def display(sp):
    slabs = slabs_stats(sp)

    print "  # Chunk_Size  Max_age   Pages   Count   Full?",
    print "  Evicted Evict_Time OOM     Used   Wasted"

    for num in sorted(slabs.keys()):
        if num == 'total':
            continue
        slab = slabs[num]
        number_els = slab['number']

        is_slab_full = "yes" if slab['free_chunks_end'] == 0 else "no"
        wasted = 0.0
        if number_els:
            wasted = float(slab['chunk_size']) * float(number_els)
            wasted = float(slab['mem_requested']) / wasted
            wasted = (1.0 - wasted) * 100

        print "{0:3d} {1:>10s} {2:7d}s {3:7d} {4:7d}" \
            " {5:>7s} {6:8d} {7:10d} {8:3d} {9:>8s}" \
            " {10:7.0f}%".format(
                num, descriptive_size(slab['chunk_size']), slab['age'],
                slab['total_pages'], slab['number'], is_slab_full,
                slab['evicted'], slab['evicted_time'], slab['outofmemory'],
                descriptive_size(slab['mem_requested']), wasted
            )

    print "\n"+"Total:"
    for prop, val in slabs['total'].iteritems():
        if prop == 'total_malloced':
            print "{0:15s} {1:>12s}".format(prop, descriptive_size(val))
        else:
            print "{0:15s} {1:>12d}".format(prop, val)

    # Calculate possible allocated memory
    stats = get_stats(sp, 'stats settings')
    item_size_max = int(stats['item_size_max'])
    maxbytes = int(stats['maxbytes'])
    pages = 1
    chunk_size = 96.0
    while chunk_size * float(stats['growth_factor']) < item_size_max:
        pages += 1
        chunk_size *= float(stats['growth_factor'])

    real_min = descriptive_size(max(item_size_max * pages, maxbytes))
    real_max = descriptive_size(
        item_size_max * (pages + maxbytes / item_size_max - 1))
    print "{0:15s} {1:>12s} (real {2} - {3})".format(
        'maxbytes', descriptive_size(maxbytes), real_min, real_max)

    # Other settings & statistic
    print "{0:15s} {1:>12s}".format(
        'item_size_max', descriptive_size(item_size_max))
    print "{0:15s} {1:>12s}".format(
        'evictions', stats['evictions'])
    print "{0:15s} {1:>12s}".format(
        'growth_factor', str(stats['growth_factor']))

    return


def sizes(sp):
    stats = get_stats(sp, 'stats settings')
    item_size_max = int(stats['item_size_max'])
    growth_factor = float(stats['growth_factor'])

    print "{0:10s} {1:>10s} {2:>10s} {3:>10s}".format(
        'Size', 'Items', 'Chunk_Size', 'Wasted')

    sizes = get_stats(sp, 'stats sizes')
    for prop in sorted([int(key) for key in sizes.keys()]):
        size = int(prop)
        val = sizes[str(prop)]

        chunk_size = 96.0
        while chunk_size * growth_factor < size:
            chunk_size *= growth_factor
        chunk_size *= growth_factor
        if chunk_size * growth_factor > item_size_max:
            chunk_size = float(item_size_max)

        wasted = (1.0 - float(size) / chunk_size) * 100

        print "{0:10s} {1:10d} {2:>10s} {3:9.0f}%".format(
            descriptive_size(size), int(val),
            descriptive_size(chunk_size), wasted)

    return


def iterkeys(sp, dumpmode=DUMPMODE_ONLYKEYS):
    slabs = slabs_stats(sp)
    stats = get_stats(sp)
    never_expire_ts = int(stats['time']) - int(stats['uptime'])

    print "      {0:40s} {1:>20s} {2:>10s} {3:>8s}".format(
        'Key', 'Expire status', 'Size', 'Waste')

    for num in sorted(slabs.keys()):
        if num == 'total':
            continue

        slab = slabs[num]
        if slab['number'] == 0:
            continue

        lines = send_and_receive(
            sp,
            "stats cachedump {0} {1}".format(str(num), str(slab['number']))
        )
        for line in lines:
            item_stat = re.match('^ITEM ([^\s]+) \[(\d+) b; (\d+) s\]', line)
            if item_stat:
                item_key, item_size, item_expiration = item_stat.groups()
                item_expiration = int(item_expiration)

                now = time.time()

                wasted = float(item_size) / float(slab['chunk_size'])
                wasted = (1.0 - wasted) * 100
                if item_expiration == never_expire_ts:
                    status = '[never expire]'
                elif now > item_expiration:
                    status = '[expired]'
                else:
                    status = str(int(item_expiration - now))+'s left'

                print "ITEM  {0:40s} {1:>20s} {2:>10s} {3:7.0f}%".format(
                    item_key, status, item_size, wasted)

                # Get expired value == remove expired value
                if dumpmode == REMOVEMODE_EXPIRED and status == '[expired]':
                    send_and_receive(sp, "get "+item_key)

                if dumpmode == DUMPMODE_KEYVALUES and status != '[expired]':
                    lines = send_and_receive(sp, "get " + item_key)
                    if len(lines):
                        item_info, item_data = lines
                        item_info_stat = re.match(
                            '^VALUE ([^\s]+) (\d+) (\d+)',
                            item_info)
                        flags = int(item_info_stat.group(2))
                        print "VALUE {0:40s} flags={1:X}".format(
                            item_key, flags)
                        print "{0:s}".format(item_data)

    return


# Default values
host = 'localhost'
port = 11211
socketpath = None
mode = 'display'

# Parse command line
if len(sys.argv) > 1:
    if sys.argv[1].find('/') == 0:
        socketpath = sys.argv[1]
        host = port = None
    elif sys.argv[1].find(':') > 0:
        host, port = sys.argv[1].split(':')
    else:
        host = sys.argv[1]
if len(sys.argv) > 2:
    mode = sys.argv[2]

# All modes
modes = ('display', 'dumpkeys', 'dump', 'removeexp', 'settings', 'stats',
         'sizes')

# Check params
if len(sys.argv) < 2 or mode not in modes:
    myhelp()
    sys.exit()


# Connect to memcached
try:
    if host and port:
        sp = socket.create_connection((host, port), DEFAULT_TCP_TIMEOUT)
    else:
        sp = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sp.connect(socketpath)
except socket.error as msg:
    print msg
    sys.exit()
else:
    if mode == 'stats':
        show_stats(sp)
    elif mode == 'settings':
        show_stats(sp, 'stats settings')
    elif mode == 'sizes':
        sizes(sp)
    elif mode == 'removeexp':
        iterkeys(sp, REMOVEMODE_EXPIRED)
    elif mode == 'dump':
        iterkeys(sp, DUMPMODE_KEYVALUES)
    elif mode == 'dumpkeys':
        iterkeys(sp, DUMPMODE_ONLYKEYS)
    else:
        display(sp)

    sp.close()


sys.exit()
